src/other/utils.jl

```diff
@@ -479,48 +479,38 @@ function _gather_groups_old_version(ds, cols, ::Val{T}; mapformats = false) wher
 end
 
 # ds assumes is grouped based on cols and groups are gathered togther
-function  _find_starts_of_groups(ds, cols::Vector, ::Val{T}; mapformats = true) where T
+function _find_starts_of_groups(ds, cols::Vector, ::Val{T}; mapformats = true) where T
     colsidx = index(ds)[cols]
 
     ranges = Vector{T}(undef, nrow(ds))
-    ranges_cpy = copy(ranges)
-    ranges[1] = 1
-    ranges_cpy[1] = 1
+    inbits = zeros(Bool, nrow(ds))
+    inbits[1] = true
     last_valid_index = 1
-
+	
     for j in 1:length(colsidx)
         if mapformats
             _f = getformat(ds, colsidx[j])
         else
             _f = identity
         end
-        last_valid_index = _find_starts_of_groups!(_columns(ds)[colsidx[j]], _f , ranges, ranges_cpy, last_valid_index)
+		_find_starts_of_groups!(_columns(ds)[colsidx[j]], _f , inbits)
+    end
+    @inbounds for i in 1:length(inbits)
+        if inbits[i] == true
+            ranges[last_valid_index] = i
+            last_valid_index += 1
+        end
     end
-    return colsidx, ranges, last_valid_index
+    return colsidx, ranges, (last_valid_index - 1)
 end
 
 _find_starts_of_groups(ds, col::ColumnIndex, ::Val{T}; mapformats = true) where T = _find_starts_of_groups(ds, [col], Val(T), mapformats = mapformats)
 _find_starts_of_groups(ds, cols::UnitRange, ::Val{T}; mapformats = true) where T = _find_starts_of_groups(ds, collect(cols), Val(T), mapformats = mapformats)
 
-
-function _find_starts_of_groups!(x, format, ranges, ranges_cpy, last_valid_index)
-    cnt = 1
-    @inbounds for j in 1:last_valid_index
-        lo = ranges_cpy[j]
-        j == last_valid_index ? hi = length(x) : hi = ranges_cpy[j + 1] - 1
-        ranges[cnt] = lo
-        cnt += 1
-        @inbounds for i in lo:(hi - 1)
-            if !isequal(format(x[i]), format(x[i+1]))
-                ranges[cnt] = i + 1
-                cnt += 1
-            end
-        end
-    end
-    @inbounds for j in 1:(cnt - 1)
-        ranges_cpy[j] = ranges[j]
+function _find_starts_of_groups!(x, f, inbits)
+    Threads.@threads for j in 2:length(inbits)
+        @inbounds inbits[j] = inbits[j]==true ? true : !isequal(f(x[j]), f(x[j-1]))
```

Comment:
why not `inbits[j] ? nothing : inbits[j] = !isequal(f(x[j]), f(x[j-1]))`

Reply:
`true` means the start of a group, `inbits[j]==true` means for at least one previous column, `f(x[j]) != f(x[j-1])`, so no need to do comparison for this `j`.

If we initialize the `inbits` using `zeros(Bool, nrow(ds))`, why there is `nothing` in the vector?

Reply:
`inbits[j] ? nothing : inbits[j] = !isequal(f(x[j]), f(x[j-1]))` actually translate to 
```julia
if !inbits[j]
     inbits[j] = !isequal(f(x[j]), f(x[j-1]))
end
```
no need to worry about `nothing`

Reply:
OK. 
When we start from `zeros(Bool, nrow(ds))`, all positions except the first row should be `false`. If `inbits[j]` is `true`, we will never change that because at least one column `xi` exists `f(xi[j]) != f(xi[j-1])`. This `j`-th row will be the start of group so no need to do `!isequal(f(x[j]), f(x[j-1]))`.

Things will be different here if grouping information of the current `ds` can be used. I will carefully check this case.